Packages/com.bci4kids.bciessentials/Runtime/Scripts/Behaviors/P300ControllerBehavior.cs

```diff
@@ -248,548 +356,941 @@ protected override IEnumerator OnStimulusRunBehavior()
                     {
                         markerString = markerString + "," + "-1";
                     }
+                    Debug.LogWarning("MARKERS ARE BEING SENT FOR OBJECT IDS NOT OBJECT POSITIONS, SO THIS WILL NOT WORK WITH BESSY PYTHON JUST YET");
+                    markerString = markerString + "," + currentObject.GetComponent<SPO>().ObjectID.ToString();
 
-                    markerString = markerString + "," + stimOrder[i].ToString();
+                    // Turn on
+                    currentObject.GetComponent<SPO>().StartStimulus();
 
-                    ///
-                    ////create and start a Stopwatch instance
-                    //Stopwatch stopwatch = Stopwatch.StartNew();
+                    // Send marker
+                    if (!blockOutGoingLSL)
+                    {
+                        marker.Write(markerString);
+                    }
 
-                    //Turn on
+                    // Wait
+                    yield return new WaitForSecondsRealtime(onTime);
 
-                    timeOfFlash = currentObject.GetComponent<SPO>().StartStimulus();
+                    // Turn off
+                    currentObject.GetComponent<SPO>().StopStimulus();
 
+                    // Wait
+                    yield return new WaitForSecondsRealtime(offTime);
+                }
+            }
+        }
 
-                    //Send marker
-                    if (blockOutGoingLSL == false)
+        
+        private IEnumerator ContextAwareMultiFlashRoutine()
+        {
+            //Total number of flashes for each grouping of objects.
+            int totalFlashes = numFlashesPerObjectPerSelection;
+                      
+            for (int jj = 0; jj < totalFlashes; jj++)
+            {
+                Debug.Log("Getting the graph partition for the Context-Aware MultiFlash, updating each loop");
+                var (subset1,subset2) = CalculateGraphPartition(_validGOs);
+
+                //Turn the subsets into randomized matrices,
+                int[,] randMat1 = SubsetToRandomMatrix(subset1);
+                int[,] randMat2 = SubsetToRandomMatrix(subset2);
+                
+                //Flash through the rows of randMat1 first, then randMat2.
+                //Off time is included in these coroutines.
+                yield return StartCoroutine(FlashRowsSubsets(randMat1));
+                yield return StartCoroutine(FlashRowsSubsets(randMat2));
+                yield return StartCoroutine(FlashColsSubsets(randMat1));
+                yield return StartCoroutine(FlashColsSubsets(randMat2));
```

Comment:
Shouldn't these be just once for randMat1 and once for randMat2? Also, maybe we could set a number of submatrices in case we want something other than just two groups?

Reply:
No. So this is written out to specifically use bipartitioning of the graph to accomplish the "set-up" for standard checkerboard flashing - https://www-sciencedirect-com.ezproxy.lib.ucalgary.ca/science/article/pii/S1388245710000738

By doing this in this way, we are first do the rows of Mat 1, then rows of Mat 2, then Cols of Mat 1 and cols of Mat 2, exactly how the implementation is done in the above paper. 

Going beyond bipartitioning is outside the scope of this work (as this is to just adapt the world to work with checkerboard pattern flashing even in 3D or other spaces)